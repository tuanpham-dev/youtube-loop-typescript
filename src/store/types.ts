export interface Video {
  id: number
  youtubeId: string
  volume: number | null
  range: [number, number] | [null, null]
}

export interface State {
  videos: Video[]
  playingVideo: number | null
}

export const ADD_VIDEO_BY_YOUTUBE_ID = 'ADD_VIDEO_BY_YOUTUBE_ID'
export const ADD_VIDEO = 'ADD_VIDEO'
export const REMOVE_VIDEO = 'REMOVE_VIDEO'
export const EDIT_VIDEO = 'EDIT_VIDEO'
export const PLAY_VIDEO = 'PLAY_VIDEO'
export const PLAY_FIRST_VIDEO = 'PLAY_FIRST_VIDEO'
export const PLAY_NEXT_VIDEO = 'PLAY_NEXT_VIDEO'
export const PLAY_PREVIOUS_VIDEO = 'PLAY_PREVIOUS_VIDEO'
export const UPDATE_VIDEOS = 'UPDATE_VIDEOS'

interface addVideoByYouTubeIdAction {
  type: typeof ADD_VIDEO_BY_YOUTUBE_ID
  payload: string
}

interface addVideoAction {
  type: typeof ADD_VIDEO
  payload: Video
}

interface removeVideoAction {
  type: typeof REMOVE_VIDEO
  payload: number
}

interface editVideoAction {
  type: typeof EDIT_VIDEO
  payload: Video
}

interface updateVideosAction {
  type: typeof UPDATE_VIDEOS
  payload: Video[]
}

interface playVideoAction {
  type: typeof PLAY_VIDEO
  payload: number
}

interface playFirstVideoAction {
  type: typeof PLAY_FIRST_VIDEO
}

interface playPreviousVideoAction {
  type: typeof PLAY_PREVIOUS_VIDEO
}

interface playNextVideoAction {
  type: typeof PLAY_NEXT_VIDEO
}

interface stopVideoAction {
  type: typeof PLAY_VIDEO
  payload: null
}

type ActionByType<T extends ActionType['type'], A = ActionType> = A extends {type: T} ? A : never

export type ActionType = addVideoByYouTubeIdAction | addVideoAction | removeVideoAction
  | editVideoAction| updateVideosAction | playVideoAction | playFirstVideoAction
  | playPreviousVideoAction | playNextVideoAction | stopVideoAction  
export type ReducerHandler<T extends ActionType['type']> = (state: State, action: ActionByType<T>) => State
export type ReducerMap = {
  [T in ActionType['type']]?: ReducerHandler<T>
}