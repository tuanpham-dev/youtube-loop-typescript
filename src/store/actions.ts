import {
  Video,
  ActionType,
  ADD_VIDEO_BY_YOUTUBE_ID,
  ADD_VIDEO, REMOVE_VIDEO,
  EDIT_VIDEO,
  UPDATE_VIDEOS,
  PLAY_VIDEO,
  PLAY_PREVIOUS_VIDEO,
  PLAY_NEXT_VIDEO,
  PLAY_FIRST_VIDEO
} from './types'
import getYouTubeVideoId from '../utils/getYouTubeVideoId'

export const addVideoByYouTubeId = (youtubeId: string): ActionType => {
  const youtubeVideoId = getYouTubeVideoId(youtubeId)

  return {
    type: ADD_VIDEO_BY_YOUTUBE_ID,
    payload: youtubeVideoId
  }
}

export const addVideo = (video: Video): ActionType => {
  return {
    type: ADD_VIDEO,
    payload: video
  }
}

export const removeVideo = (videoId: number): ActionType => {
  return {
    type: REMOVE_VIDEO,
    payload: videoId
  }
}

export const editVideo = (video: Video): ActionType => {
  return {
    type: EDIT_VIDEO,
    payload: video
  }
}

export const updateVideos = (videos: Video[]): ActionType => {
  return {
    type: UPDATE_VIDEOS,
    payload: videos
  }
}

export const playVideo = (videoId: number): ActionType => {
  return {
    type: PLAY_VIDEO,
    payload: videoId
  }
}

export const playFirstVideo = (): ActionType => {
  return {
    type: PLAY_FIRST_VIDEO
  }
}

export const playPreviousVideo = (): ActionType => {
  return {
    type: PLAY_PREVIOUS_VIDEO
  }
}

export const playNextVideo = (): ActionType => {
  return {
    type: PLAY_NEXT_VIDEO
  }
}

export const stopVideo = (): ActionType => {
  return {
    type: PLAY_VIDEO,
    payload: null
  }
}
