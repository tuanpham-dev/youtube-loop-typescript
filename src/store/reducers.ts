import {
  Video,
  State,
  ActionType,
  ReducerMap,
  ADD_VIDEO_BY_YOUTUBE_ID,
  ADD_VIDEO,
  REMOVE_VIDEO,
  EDIT_VIDEO,
  UPDATE_VIDEOS,
  PLAY_VIDEO,
  PLAY_FIRST_VIDEO,
  PLAY_PREVIOUS_VIDEO,
  PLAY_NEXT_VIDEO,
  ReducerHandler
} from "./types";

const initialState: State = {
  videos: [],
  playingVideo: null
}

const reducerMap: ReducerMap = {
  [ADD_VIDEO_BY_YOUTUBE_ID]: (state, { payload: youtubeId }) => {
    const maxId = Math.max(...state.videos.map(video => video.id), 0)
    const video: Video = {
      id: maxId + 1,
      youtubeId,
      volume: null,
      range: [null, null]
    }

    return {
      ...state,
      videos: [...state.videos, video]
    }
  },

  [ADD_VIDEO]: (state, { payload: video }) => {
    const maxId = Math.max(...state.videos.map(video => video.id), 0)

    return {
      ...state,
      videos: [...state.videos, { ...video, id: maxId + 1 }]
    }
  },

  [REMOVE_VIDEO]: (state, { payload: videoId }) => {
    const index = state.videos.findIndex(video => video.id === videoId)
    let playingVideo: number | null = null

    if (index !== null) {
      if (state.playingVideo === videoId) {
        if (state.videos.length <= 1) {
          playingVideo = null
        } else {
          playingVideo = state.videos[index === state.videos.length - 1 ? 0 : index + 1].id
        }
      }

      const videos = [...state.videos]
      videos.splice(index, 1)

      if (playingVideo !== null) {
        return {
          ...state,
          videos,
          playingVideo
        }
      } else {
        return {
          ...state,
          videos
        }
      }
    }

    return state
  },

  [EDIT_VIDEO]: (state, { payload: video }) => {
    const index = state.videos.findIndex(videoItem => videoItem.id === video.id)

    if (index !== null) {
      const videos = [...state.videos]
      videos.splice(index, 1, video)

      return {
        ...state,
        videos
      }
    }

    return state
  },

  [UPDATE_VIDEOS]: (state, { payload: videos }) => {
    return {
      ...state,
      videos
    }
  },

  [PLAY_VIDEO]: (state, { payload: videoId }) => {
    if (state.playingVideo !== videoId) {
      return {
        ...state,
        playingVideo: videoId
      }
    }

    return state
  },

  [PLAY_FIRST_VIDEO]: (state) => {
    if (state.videos.length > 0) {
      return {
        ...state,
        playingVideo: state.videos[0].id
      }
    }

    return state
  },

  [PLAY_PREVIOUS_VIDEO]: (state) => {
    if (state.videos.length > 0) {
      let previousVideoIndex
      const index = state.videos.findIndex(video => video.id === state.playingVideo)
  
      if (index === 0) {
        previousVideoIndex = state.videos.length - 1
      } else {
        previousVideoIndex = index -1
      }

      return {
        ...state,
        playingVideo: state.videos[previousVideoIndex].id
      }
    }

    return state
  },

  [PLAY_NEXT_VIDEO]: (state) => {
    if (state.videos.length > 0) {
      let nextVideoIndex
      const index = state.videos.findIndex(video => video.id === state.playingVideo)

      if (index === state.videos.length - 1) {
        nextVideoIndex = 0
      } else {
        nextVideoIndex = index + 1
      }

      return {
        ...state,
        playingVideo: state.videos[nextVideoIndex].id
      }
    }

    return state
  }
}

export const rootReducer = (state = initialState, action: ActionType): State => {
  if (typeof reducerMap[action.type] === 'function') {
    const reducerHandler = reducerMap[action.type] as ReducerHandler<typeof action.type>

    return reducerHandler(state, action)
  } else {
    return state
  }
}

export type RootState = ReturnType<typeof rootReducer>
export default rootReducer