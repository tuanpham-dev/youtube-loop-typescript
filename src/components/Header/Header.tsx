import React, { FunctionComponent, useState } from 'react'
import { connect, ConnectedProps } from 'react-redux'
import { State } from '../../store/types'
import Button from '../Button/Button'
import { addVideoByYouTubeId, stopVideo, playNextVideo, playPreviousVideo, playFirstVideo } from '../../store/actions'
import { Dispatch } from 'redux'

const mapStateToProps = (state: State) => ({
  videos: state.videos,
  playingVideo: state.playingVideo
})

const mapDispatchToProps = (dispatch: Dispatch) => ({
  onAddButtonClick: (input: string) => {
    dispatch(addVideoByYouTubeId(input))
  },
  onPlayButtonClick: () => {
    dispatch(playFirstVideo())
  },
  onStopButtonClick: () => {
    dispatch(stopVideo())
  },
  onNextButtonClick: () => {
    dispatch(playNextVideo())
  },
  onPreviousButtonClick: () => {
    dispatch(playPreviousVideo())
  }
})

const connector = connect(mapStateToProps, mapDispatchToProps)

type PropsFromRedux = ConnectedProps<typeof connector>
type HeaderProps = PropsFromRedux

export const Header: FunctionComponent<HeaderProps> = ({ videos, playingVideo, onAddButtonClick, onPlayButtonClick, onStopButtonClick, onPreviousButtonClick, onNextButtonClick }) => {
  const [input, setInput] = useState('')

  const videosCount = videos.length
  const isPlaying = playingVideo !== null
  const canPlayPrevNext = isPlaying && videosCount > 1

  return (
    <div>
      <header className="header">
        <div className="container">
          <a href="/" className="header__logo">YouTube Loop</a>

          {videosCount > 0 &&
            <div className="header__controls">
              {isPlaying?
                <Button onClick={onStopButtonClick}>Stop</Button>
                :
                <Button onClick={onPlayButtonClick}>Play</Button>
              }

              {canPlayPrevNext &&
                <>
                  <Button onClick={onPreviousButtonClick}>Previous</Button>
                  <Button onClick={onNextButtonClick}>Next</Button>
                </>
              }
            </div>
          }

          <div className="header__input-group">
            <input type="text" className="header__input"
              placeholder="Enter YouTube URL or Video ID"
              value={input}
              onChange={event => setInput(event.target.value)}
            />
            <div className="header__input-group-append">
              <Button color="blue" onClick={() => onAddButtonClick(input)}>Go Loop!</Button>
            </div>
          </div>
        </div>
      </header>
    </div>
  )
}

export default connector(Header)
