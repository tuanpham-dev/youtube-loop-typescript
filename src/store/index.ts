import { Store, createStore } from 'redux'
import { composeWithDevTools } from 'redux-devtools-extension'
import { rootReducer, RootState } from './reducers'

export const configureStore = (initialState?: RootState): Store<RootState> => {
  const middleware = composeWithDevTools()
  const store = createStore(rootReducer, initialState, middleware) as Store<RootState>

  return store
}

export default configureStore