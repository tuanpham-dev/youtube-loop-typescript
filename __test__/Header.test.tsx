import React from 'react'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import TestRenderer from 'react-test-renderer'
import Header from '../src/components/Header/Header'
import mockData from './mock-data/video'
import { addVideoByYouTubeId, playFirstVideo, stopVideo, playPreviousVideo, playNextVideo } from '../src/store/actions'

const mockStore = configureMockStore()
const initialState = {
  videos: [],
  playingVideo: null
}

describe('Header', () => {
  describe('Renderer', () => {
    it('should render with no control button', () => {
      const store = mockStore(initialState)
      const header = TestRenderer.create(
        <Provider store={store}>
          <Header />
        </Provider>
      )
      expect(header.toJSON()).toMatchSnapshot()
    })

    it('should render with play button', () => {
      const store = mockStore({
        ...initialState,
        videos: mockData.videos
      })
      const header = TestRenderer.create(
        <Provider store={store}>
          <Header />
        </Provider>
      )
      expect(header.toJSON()).toMatchSnapshot()
    })

    it('should render with stop button', () => {
      const store = mockStore({
        videos: [mockData.video],
        playingVideo: mockData.video.id
      })
      const header = TestRenderer.create(
        <Provider store={store}>
          <Header />
        </Provider>
      )
      expect(header.toJSON()).toMatchSnapshot()
    })

    it('should render with stop, previous and next buttons', () => {
      const store = mockStore({
        videos: mockData.videos,
        playingVideo: mockData.videos[2].id
      })
      const header = TestRenderer.create(
        <Provider store={store}>
          <Header />
        </Provider>
      )
      expect(header.toJSON()).toMatchSnapshot()
    })
  })

  describe('Actions', () => {
    it('should dispatch addVideoByYouTubeId on Go Loop! button click', () => {
      const store = mockStore(initialState)
      store.dispatch = jest.fn();
      const header = TestRenderer.create(
        <Provider store={store}>
          <Header />
        </Provider>
      )

      const buttons = header.root.findAllByType('button')
      expect(buttons[0].children).toEqual(['Go Loop!'])

      TestRenderer.act(() => {
        header.root.findByType('input').props.onChange({ target: { value: mockData.youtubeId }})
      })

      TestRenderer.act(() => {
        header.root.findByType('button').props.onClick()
      })

      expect(store.dispatch).toHaveBeenCalledTimes(1)
      expect(store.dispatch).toHaveBeenCalledWith(
        addVideoByYouTubeId(mockData.youtubeId)
      )
    })

    it('should dispatch playFirstVideo on Play button click', () => {
      const store = mockStore({
        ...initialState,
        videos: mockData.videos
      })
      store.dispatch = jest.fn();
      const header = TestRenderer.create(
        <Provider store={store}>
          <Header />
        </Provider>
      )

      const buttons = header.root.findAllByType('button')
      expect(buttons[0].children).toEqual(['Play'])

      TestRenderer.act(() => {
        header.root.findAllByType('button')[0].props.onClick()
      })

      expect(store.dispatch).toHaveBeenCalledTimes(1)
      expect(store.dispatch).toHaveBeenCalledWith(
        playFirstVideo()
      )
    })

    it('should dispatch stopVideo on Stop button click', () => {
      const store = mockStore({
        videos: mockData.videos,
        playingVideo: mockData.videos[2].id
      })
      store.dispatch = jest.fn();
      const header = TestRenderer.create(
        <Provider store={store}>
          <Header />
        </Provider>
      )

      const buttons = header.root.findAllByType('button')
      expect(buttons[0].children).toEqual(['Stop'])

      TestRenderer.act(() => {
        header.root.findAllByType('button')[0].props.onClick()
      })

      expect(store.dispatch).toHaveBeenCalledTimes(1)
      expect(store.dispatch).toHaveBeenCalledWith(
        stopVideo()
      )
    })

    it('should dispatch playPreviousVideo on Previous button click', () => {
      const store = mockStore({
        videos: mockData.videos,
        playingVideo: mockData.videos[2].id
      })
      store.dispatch = jest.fn();
      const header = TestRenderer.create(
        <Provider store={store}>
          <Header />
        </Provider>
      )

      const buttons = header.root.findAllByType('button')
      expect(buttons[1].children).toEqual(['Previous'])

      TestRenderer.act(() => {
        header.root.findAllByType('button')[1].props.onClick()
      })

      expect(store.dispatch).toHaveBeenCalledTimes(1)
      expect(store.dispatch).toHaveBeenCalledWith(
        playPreviousVideo()
      )
    })

    it('should dispatch playNextVideo on Next button click', () => {
      const store = mockStore({
        videos: mockData.videos,
        playingVideo: mockData.videos[2].id
      })
      store.dispatch = jest.fn();
      const header = TestRenderer.create(
        <Provider store={store}>
          <Header />
        </Provider>
      )

      const buttons = header.root.findAllByType('button')
      expect(buttons[2].children).toEqual(['Next'])

      TestRenderer.act(() => {
        header.root.findAllByType('button')[2].props.onClick()
      })

      expect(store.dispatch).toHaveBeenCalledTimes(1)
      expect(store.dispatch).toHaveBeenCalledWith(
        playNextVideo()
      )
    })
  })
})