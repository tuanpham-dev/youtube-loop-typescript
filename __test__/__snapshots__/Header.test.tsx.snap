// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Renderer should render with no control button 1`] = `
<div>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <a
        className="header__logo"
        href="/"
      >
        YouTube Loop
      </a>
      <div
        className="header__input-group"
      >
        <input
          className="header__input"
          onChange={[Function]}
          placeholder="Enter YouTube URL or Video ID"
          type="text"
          value=""
        />
        <div
          className="header__input-group-append"
        >
          <button
            className="button button--blue"
            onClick={[Function]}
          >
            Go Loop!
          </button>
        </div>
      </div>
    </div>
  </header>
</div>
`;

exports[`Header Renderer should render with play button 1`] = `
<div>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <a
        className="header__logo"
        href="/"
      >
        YouTube Loop
      </a>
      <div
        className="header__controls"
      >
        <button
          className="button"
          onClick={[Function]}
        >
          Play
        </button>
      </div>
      <div
        className="header__input-group"
      >
        <input
          className="header__input"
          onChange={[Function]}
          placeholder="Enter YouTube URL or Video ID"
          type="text"
          value=""
        />
        <div
          className="header__input-group-append"
        >
          <button
            className="button button--blue"
            onClick={[Function]}
          >
            Go Loop!
          </button>
        </div>
      </div>
    </div>
  </header>
</div>
`;

exports[`Header Renderer should render with stop button 1`] = `
<div>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <a
        className="header__logo"
        href="/"
      >
        YouTube Loop
      </a>
      <div
        className="header__controls"
      >
        <button
          className="button"
          onClick={[Function]}
        >
          Stop
        </button>
      </div>
      <div
        className="header__input-group"
      >
        <input
          className="header__input"
          onChange={[Function]}
          placeholder="Enter YouTube URL or Video ID"
          type="text"
          value=""
        />
        <div
          className="header__input-group-append"
        >
          <button
            className="button button--blue"
            onClick={[Function]}
          >
            Go Loop!
          </button>
        </div>
      </div>
    </div>
  </header>
</div>
`;

exports[`Header Renderer should render with stop, previous and next buttons 1`] = `
<div>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <a
        className="header__logo"
        href="/"
      >
        YouTube Loop
      </a>
      <div
        className="header__controls"
      >
        <button
          className="button"
          onClick={[Function]}
        >
          Stop
        </button>
        <button
          className="button"
          onClick={[Function]}
        >
          Previous
        </button>
        <button
          className="button"
          onClick={[Function]}
        >
          Next
        </button>
      </div>
      <div
        className="header__input-group"
      >
        <input
          className="header__input"
          onChange={[Function]}
          placeholder="Enter YouTube URL or Video ID"
          type="text"
          value=""
        />
        <div
          className="header__input-group-append"
        >
          <button
            className="button button--blue"
            onClick={[Function]}
          >
            Go Loop!
          </button>
        </div>
      </div>
    </div>
  </header>
</div>
`;
